### DISCUSSION TITLE
ncplane_rgba and rgba values vs ncpixels

### DISCUSSION THREAD

**Discussion by joseluis** [dankamongmen/notcurses#1237] [General]
ncplane_rgba writes an rgba array, but that array is not of ncpixels, right? Since internally an ncpixel is **rbga**. So there are 2 different *rgba* types used in the library.

**Comment by dankamongmen** [comment]
brother, don't i know it :(. it has been the source of much confusion. there are actually kinda three: rgba, rbga, and `nccell`/channels, which are rgb+{2 bits of A + some crap}. the rbga business arises from endianness issues.

i hadn't considered `ncplane_rgba()` though. if that's going to rgba, it should probably be changed to go to rbga. or, to maintain backwards compat, we should just introduce `ncplane_rbga()`.

BUT, these are subtle and delicate matters, and i want to think about it. i thought `ncpixel` was actually ABGR, for instance? to be honest i've got a bit of a sunday drunk on and am not at my sharpest regarding pixel formats at the moment. i'll return to this. thanks for bringing it up.

**Reply by joseluis** [reply]
> i hadn't considered `ncplane_rgba()` though. if that's going to rgba, it should probably be changed to go to rbga. or, to maintain backwards compat, we should just introduce `ncplane_rbga()`.

I also think it would be a good idea to use less different color types if there's not performance hit for doing so. 

> BUT, these are subtle and delicate matters, and i want to think about it. i thought `ncpixel` was actually ABGR, for instance?

From what I learned from porting the [ncpixel functions](https://github.com/dankamongmen/notcurses/blob/master/rust/src/pixel.rs) it appears to be a AGBR (or RBGA from LSB).

**Reply by joseluis** [reply]
I thought I was going crazy, but now I undestand what's happening..

Among the very first things I did on the rust bindings was to port the ncpixels_* static functions, which I have't touched since then.

These are the ncpixel_g() and ncpixel_b() functions  from a few months ago: 
```
static inline int
ncpixel_g(uint32_t pixel){
  return (pixel & 0x00ff0000ul) >> 16u;
}

static inline int
ncpixel_b(uint32_t pixel){
  return (pixel & 0x0000ff00ul) >> 8u;
}
```

And here is how they're now:
```
// Extract the 8-bit green component from an ABGR pixel
static inline unsigned
ncpixel_g(uint32_t pixel){
  return (htole(pixel) & 0x0000ff00ul) >> 8u;
}

// Extract the 8-bit blue component from an ABGR pixel
static inline unsigned
ncpixel_b(uint32_t pixel){
  return (htole(pixel) & 0x00ff0000ul) >> 16u;
}
```

Green and blue bit positions were reversed at some point.... These are the kind of things I feared, unnoticed changes to the static inline functions that are not propagated to the bindings...

**Reply by dankamongmen** [reply]
can you refresh me on why you need reimplement the inline functions? is it because bindgen can't compile them down, basically?

**Reply by joseluis** [reply]
Yeah, that was the reason.

I was looking on how to proceed with updating the functions, and specifically the `htole` macro which depends on different c headers... I want to avoid depending on more crates. Probably [to_le](https://doc.rust-lang.org/std/primitive.u32.html#method.to_le) and [to_be](https://doc.rust-lang.org/std/primitive.u32.html#method.to_be) u32 methods are enough. I'll take a stab at it. In the worst case I see we have already the [byteorder crate](https://crates.io/crates/byteorder) already in debian and fedora..

**Reply by joseluis** [reply]
Since I don't usually work with endianness, and to avoid silly mistakes, I'd like to check first with you:

```
// take host byte order and turn it into network (reverse on LE, no-op on BE),
// then reverse that, guaranteeing LE. htole(x) == ltohe(x).
```

Are all of this operations necessary? or do we just need to guarantee a LE order? i.e. [to_le](https://doc.rust-lang.org/std/primitive.u32.html#method.to_le)

**Reply by dankamongmen** [reply]
be aware that i'm hoping to get rid of this code, as it's (a) gross and (b) broken for dynamic-endian machines, when compiled under one endianness and executed on another. i just haven't figured out the best way to do so yet.

we've unfortunately got to deal with endianness at some point, since we're both treating that field as an integer sometimes, and a char* sometimes. :/

**Reply by dankamongmen** [reply]
also, what we have there can *definitely* be simplified using the `host_to_leXX()` functionality of linux's `byteorder.h`, though i need find freebsd equivalents to do so (i might roll our own).

**Reply by joseluis** [reply]
If the `to_le` method does the same as your `htole` macro then it should be working righ [now](https://github.com/dankamongmen/notcurses/blob/master/rust/src/pixel.rs#L66). I don't know yet how to test it though.

**Comment by dankamongmen** [comment]
fwiw, all RGBA etc is forced on us from external sources (ffmpeg and friends), since our native format is `nccell`.
